src/nrniv/nrnbbcore_write.cpp

```diff
@@ -1781,6 +1807,61 @@ int* datum2int(int type, Memb_list* ml, NrnThread& nt, CellGroup& cg, DatumIndic
 }
 
 
+/** @brief Return location for CoreNEURON to copy data into.
+ *  The type is mechanism type or special negative type for voltage,
+ *  i_membrane_, or time. See coreneuron/io/nrn_setup.cpp:stdindex2ptr.
+ *  We allow coreneuron to copy to NEURON's AoS data as CoreNEURON knows
+ *  how its data is arranged (SoA and possibly permuted).
+ *  This function figures out the size (just for sanity check)
+ *  and data pointer to be returned based on type and thread id.
+ *  The ARTIFICIAL_CELL type case is special as there is no thread specific
+ *  Memb_list for those.
+ */
+size_t nrnthreads_type_return(int type, int tid, double*& data, double**& mdata) {
+  size_t n = 0;
+  data = NULL;
+  mdata = NULL;
+  if (tid >= nrn_nthread) {
+    return n;
+  }
+  NrnThread& nt = nrn_threads[tid];
+  if (type == voltage) {
+    data = nt._actual_v;
+    n = size_t(nt.end);
+  } else if (type == i_membrane_) { // i_membrane_
+    data = nt._nrn_fast_imem->_nrn_sav_rhs;
+    n = size_t(nt.end);
+  } else if (type == 0) { // time
+    data = &nt._t;
+    n = 1;
+  } else if (type > 0 && type < n_memb_func) {
+    Memb_list* ml = nt._ml_list[type];
+    if (ml) {
+      mdata = ml->data;
+      n = ml->nodecount;
+    }else{
+      // The single thread case is easy
+      if (nrn_nthread == 1) {
+        ml = memb_list + type;
+        mdata = ml->data;
+        n = ml->nodecount;
+      }else{
+        // mk_tml_with_art() created a cgs[id].mlwithart which appended
+        // artificial cells to the end. Turns out that
+        // cellgroups_[tid].type2ml[type]
+        // is the Memb_list we need. Sadly, by the time we get here, cellgroups_
+        // has already been deleted.  So we defer deletion of the necessary
+        // cellgroups_ portion (deleting it on return from nrncore_run).
+        auto& p = deferred_type2artdata[tid][type];
+        n = size_t(p.first);
+        mdata = p.second;
+      }
+    }
+  }
```

Comment:
Sorry. I thought std::map would raise an error if one asked for a non-existent item. What is p in this case? Or n or mdata? Also CoreNEURON will raise an error if the 'n' return value does not match the size of the CoreNEURON array.

Reply:
> Sorry. I thought std::map would raise an error if one asked for a non-existent item. 

I was referring to `else` block corresponding to `} else if (type > 0 && type < n_memb_func) {` i.e. in case coreneuron pass `type` arg grater than `n_memb_func`.

Reply:
Ok. I see what you mean.

Reply:
Actually, that case will return n=0 and null values for data and mdata. So CoreNEURON is the place where the error will be raised (assuming it wanted anything back)